proposals/0002-spatial-navigation.md

```diff
@@ -0,0 +1,143 @@
+- Collect every focusable component and incorporate them into the SpatialNavigation instance.
+- Focus first available navigable candidate.
+- Once an arrow is pressed we should go through Basic Spatial Navigation Heuristics (call navigate(direction))
+  - Find Candidates
+  - Select the best candidate
+  - Focus the best candidate
+
+### In terms of changes to the code:
+
+We need to make sure that events can still move upwards when spatial navigation is turned on (`enableRCUNavigation` is set to true). Currently, it’s blocking all keydown events with `event.stopPropagation()`, except for TAB key.
+Block the UP and DOWN keys to execute `stepForward()` and `stepBack()` functions, as these keys will be needed for navigation from and to the progress bar (Slider).
+
+### In terms of adding the code
+
+Spatial-navigation interface and Component class extension
+
+```
+interface Positions {
+boundingClientRect: DOMRectReadOnly;
+center: DOMRectReadOnly;
+}
+
+interface Component {
+// Is considered as focusable component.
+getIsFocusable(): boolean;
+// Is currently visible/enabled/etc...
+getIsAvailableToBeFocused(): boolean;
+// on focus handler
+handleFocus(): void;
+// on blur handler
+handleBlur(): void;
+// on submit handler
+handleEnter(): void;
+// component's element boundingClientRect and center
+getPositions(): Positions;
+// focus component's element
+focus(): void;
+}
+
+enum SpatialNavigationDirection {
+Left = 'left',
+Right = 'right',
+Up = 'up',
+Down = 'down'
+}
+
+interface SpatialNavigation {
+// current focusable components:
+getComponents(): Set<Component>;
+// start list keydown events
+start(): void;
+// stop listen key down events
+stop(): void;
+// add focusable component
+add(component: Component): void;
+// remove focusable component
+remove(component: Component): void;
+// temporary pause spatial navigation
+pause(): void;
+// resume spatial navigation
+resume(): void;
+// clear current list of focusable components
+clear(): void;
+// run spatial navigation Heuristics (re-use https://github.com/WICG/spatial-navigation/blob/main/polyfill/spatial-navigation-polyfill.js#L147)
+move(direction: SpatialNavigationDirection): void;
+}
+```
+
+### In terms of changes the UI
+- Since LEFT and RIGHT keys are reserved for video seeking, we’ll have to change the UI and move the progress bar on top of the buttons in order to allow the users to use those keys to navigate between elements when buttons are focused and seek the video when the progress bar is focused
+- Current player controls are too small for big screens, so we’re aiming to make those bigger
+- Vjs-control-bar is using “display: flex;”, in order to make spatial-navigation-polyfill work properly we have to add a “gap: 1px;”
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- There might be some performance issues, especially with lower-end devices, due to the polyfill usage.
+- There might also be compatibility issues with certain platforms, especially older ones.
```

Comment:
Our minimum browser engine support in Video.js 8 is Chromium 53. Of course, that doesn't help with older devices and their hardware limitations.

Reply:
Thanks for pointing this out. I'll remove these drawbacks as they are related to the spatial-navigation polyfill which will not be included.

Reply:
Changed in the latest commit